scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,622 @@
+///   a channel that no longer has a receiver.
+///
+/// # Examples
+///
+/// ```
+/// use scipio::{LocalExecutor, Local, Latency, Shares};
+/// use scipio::channels::LocalChannel;
+/// use futures::stream::StreamExt;
+///
+/// let ex = LocalExecutor::make_default();
+/// ex.run(async move {
+///     let task_queue =
+///         Local::create_task_queue(
+///             Shares::default(), Latency::NotImportant, "example");
+///     
+///     let (sender, mut receiver) = LocalChannel::new_unbounded();
+///     let h = Local::local_into(async move {
+///         assert_eq!(receiver.next().await.unwrap(), 0);
+///     }, task_queue).unwrap().detach();
+///     sender.try_send(0);
+///     drop(sender);
+///     h.await;
+/// });
+/// ```
+/// [`TaskQueue`]: struct.TaskQueue.html
+/// [`LocalSender`]: struct.LocalSender.html
+/// [`LocalReceiver`]: struct.LocalReceiver.html
+/// [`ChannelError`]: struct.ChannelError.html
+/// [`push`]: struct.LocalSender.html#method.push
+/// [`None`]: https://doc.rust-lang.org/std/option/enum.Option.html#variant.None
+/// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+/// [`channel`]: https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html
+/// [`next`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.next
+/// [`StreamExt`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html
+pub struct LocalChannel<T> {
+    state: Rc<RefCell<State<T>>>,
+}
+
+impl<T> LocalChannel<T> {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(capacity: ChannelCapacity) -> (LocalSender<T>, LocalReceiver<T>) {
+        let channel = match capacity {
+            ChannelCapacity::Unbounded => VecDeque::new(),
+            ChannelCapacity::Bounded(x) => VecDeque::with_capacity(x),
+        };
+
+        let state = Rc::new(RefCell::new(State {
+            capacity,
+            channel,
+            send_waiters: Vec::new(),
+            recv_waiters: Vec::new(),
+            receiver_alive: true,
+            sender_alive: true,
+        }));
+
+        (
+            LocalSender {
+                channel: LocalChannel {
+                    state: state.clone(),
+                },
+            },
+            LocalReceiver {
+                channel: LocalChannel { state },
+            },
+        )
+    }
+
+    fn push(&self, item: T) -> Result<(), ChannelError<T>> {
+        let mut state = self.state.borrow_mut();
+        if !state.receiver_alive {
+            return Err(ChannelError::new(io::ErrorKind::BrokenPipe, item));
+        }
+
+        if let ChannelCapacity::Bounded(x) = state.capacity {
+            if state.channel.len() >= x {
+                return Err(ChannelError::new(io::ErrorKind::WouldBlock, item));
+            }
+        }
+
+        state.channel.push_back(item);
+        if !state.recv_waiters.is_empty() {
+            let waiters = std::mem::replace(&mut state.recv_waiters, Vec::new());
```

Comment:
The shorter way to spell this would be 

```suggestion
            let waiters = std::mem::take(&mut state.recv_waiters);
```